tests/installation/select_patterns_and_packages.pm

```diff
@@ -199,6 +203,8 @@ sub run {
                     if (match_has_tag("pattern-$p")) {
                         $needs_to_be_selected = $sel;
                         $current_pattern      = $p;
+                        # add pattern to the set of detected patterns
+                        $processed_patterns{$p} = undef;
```

Comment:
I am curious to know if you are using hashes instead of lists because is more convenient in Perl. At the end you got a list but in your processing you use hashes. Perhaps is for not using grep and use exist that might be more readable? Just wondering why don't use list all the time if you are not going to use the value in the hashes? The code is nice anyway.

Reply:
There are implementations of sets, but that would require additional cpan dependency, so using hashes as a structure for sets. Using `undef` uses less memory to store in comparison to `1`. Lists would be harder to compare basically, as not sorted, and even more complex in case of duplicates. So the loop which I have later would be O(n^2) instead of linear.

Using hashes for storing sets is common approach in other languages too, e.g. in java the is no implementation for concurrent set, as Concurrent Hash Map can be used instead.

Reply:
it's interesting this approach, in other languages similar to Java I think there is big use of an intermediate library/framework that allow you to access the element with a handy function so it is kind of more readable to use it for a few element but it is true that if you want efficiency you need similar object like you mentioned that wrap this behavior. It is nice that Perl uses this approach even when we are not using cpan additional deps.

Reply:
https://metacpan.org/pod/Set::Scalar this implementation could provide me the diff of two sets with single call, so yes, I kind of reinvented the wheel =)